### PR TITLE
[null-aggr] Add null handling support in `mode` aggregation

### PR DESCRIPTION
**Pull request by gortiz**
This PR adds null handling support in the `mode` aggregation. Specifically, it modifies the code to ignore null values when `mode` is evaluated with `nullHandlingEnabled`, trying to follow the following postgres logic:

```psql
postgres=# create table myTable(myInt INT);
CREATE TABLE
postgres=# insert into myTable(myInt) VALUES (null), (1), (null);
INSERT 0 3
postgres=# SELECT mode() WITHIN GROUP (ORDER BY myInt) AS mode FROM myTable;
 mode 
------
    1
(1 row)

```

It is recommended to review this PR assuming PR #12226 is merged, which is equivalent to compare this PR from commit commit f966b1b295134a08a3b201dadcedd91ca7d01c0a to head. As you can see there, this PR also changes the test and shows that when null handling is enabled, nulls are ignored.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+import java.util.function.DoubleConsumer;
+import java.util.function.Function;
+import java.util.function.LongConsumer;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.core.common.BlockValSet;
+import org.roaringbitmap.IntConsumer;
+import org.roaringbitmap.RoaringBitmap;
+
+
+public abstract class NullableSingleInputAggregationFunction<I, F extends Comparable>
```
Comment (walterddr) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:33]: iiuc this class would be extensible from all null-handling aggregate features? or do we plan to only use this for mode and not the other ones?

Comment (gortiz) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:33]: TL;DR: That class should be used by all aggregations once we migrate them.

I started adapting most (if not all) aggregation functions, but the PR was huge and I didn't write the test. So I decided to freeze that code. I pushed my progress to a personal branch in case I need to consult it (you can read it in https://github.com/gortiz/pinot/tree/null-aggr-bak) and I'm bringing changes from that branch into function specific branches.

Comment (gortiz) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:36]: I've moved this class to https://github.com/apache/pinot/pull/12215, which is going to be the base PR of this family of PRs

Comment (gortiz) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:36]: Finally I decided to move this code to https://github.com/apache/pinot/pull/12236, given #12215 is focused on the fluent framework itself. Anyway, both #12215 and #12236 should be merged (or discarded) soon.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+  protected final boolean _nullHandlingEnabled;
+
+  public NullableSingleInputAggregationFunction(ExpressionContext expression, boolean nullHandlingEnabled) {
+    super(expression);
+    _nullHandlingEnabled = nullHandlingEnabled;
+  }
+
+
+  @FunctionalInterface
+  public interface BatchConsumer {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:47]: I can see them simplify the code, but from past experience, the functional interface can significant impact the performance (at least for java 8). Aggregation function is one of the most performance sensitive code, and these function are applied on a per value basis. Can we do some micro benchmark to ensure there is no overhead?

Comment (gortiz) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:47]: I've created a benchmark in another PR (marked as draft). See https://github.com/apache/pinot/pull/12354

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+import java.util.function.LongConsumer;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.core.common.BlockValSet;
+import org.roaringbitmap.IntConsumer;
+import org.roaringbitmap.IntIterator;
+import org.roaringbitmap.RoaringBitmap;
+
+
+public abstract class NullableSingleInputAggregationFunction<I, F extends Comparable>
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:36]: Are we going to make all aggregation able to take `null` values? If so, maybe consider adding the new methods directly to `BaseSingleInputAggregationFunction`

Comment (gortiz) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:36]: The plan is to eventually add it to all operations when it makes sense. In the medium term the idea is to support null handling in all operations that are not multi valued or not super pinot specific, so we can increase the compatibility with _standard_ SQL

**Comment by gortiz** [comment]
@Jackie-Jiang this should be ready to merge. Can you take a look? As far as I can see the only discussion pending is whether the `NullableSingleInputAggregationFunction` should be merged with `BaseSingleInputAggregationFunction`

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java [hunk 4]
```diff
@@ -263,11 +264,15 @@ public void aggregate(int length, AggregationResultHolder aggregationResultHolde
     // For dictionary-encoded expression, store dictionary ids into the dictId map
     Dictionary dictionary = blockValSet.getDictionary();
     if (dictionary != null) {
-      int[] dictIds = blockValSet.getDictionaryIdsSV();
+
       Int2IntOpenHashMap dictIdValueMap = getDictIdCountMap(aggregationResultHolder, dictionary);
-      for (int i = 0; i < length; i++) {
-        dictIdValueMap.merge(dictIds[i], 1, Integer::sum);
-      }
+      int[] dids = blockValSet.getDictionaryIdsSV();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:269]: (nit) Shall we keep the name as `dictIds` and `dictId`? That is used across all the engine code

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java [hunk 5]
```diff
@@ -263,11 +264,15 @@ public void aggregate(int length, AggregationResultHolder aggregationResultHolde
...
+      forEachNotNull(length, blockValSet, (from, to) -> {
+        for (int i = from; i < to; i++) {
+          int did = dids[i];
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:272]: (nit) inline

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+     * @param it1 it has to iterate in ascending order and the min value is 0
+     * @param it2 it has to iterate in ascending order and the min value is 0
+     */
+    public MinIntIterator(IntIterator it1, IntIterator it2) {
+      _it1 = it1;
+      _it2 = it2;
+    }
+
+    @Override
+    public IntIterator clone() {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:207]: Do we need this override? Seems strange to me as we should never clone it

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    if (prev < length) {
+      acum = reducer.apply(acum, prev, length);
+    }
+    return acum;
+  }
+
+  /**
+   * Returns an iterator that returns the union of the null bitmaps of valSet1 and valSet2.
+   */
+  public IntIterator orNullIterator(BlockValSet valSet1, BlockValSet valSet2) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:146]: Is this used? Trying to understand where this method fits

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java [hunk 8]
```diff
@@ -467,7 +501,11 @@ public ColumnDataType getFinalResultColumnType() {
   @Override
   public Double extractFinalResult(Map<? extends Number, Long> intermediateResult) {
     if (intermediateResult.isEmpty()) {
-      return DEFAULT_FINAL_RESULT;
+      if (_nullHandlingEnabled) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:504]: When `_nullHandlingEnabled` is `true`, and all input values are `null`, will this map be `null`?
Same for `merge()`

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java [hunk 9]
```diff
@@ -254,6 +256,24 @@ public Builder withTierOverwrites(JsonNode tierOverwrites) {
       return this;
     }
 
+    public Builder withIndex(String indexId, IndexConfig indexConfig) {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java:259]: Is this related?

**Review by Jackie-Jiang — COMMENTED** [review]
Mostly good. I saw you added mode into the `AllNullQueriesTest`, but somehow I didn't find where the `null` intermediate result is handled

**Comment by gortiz** [comment]
> Mostly good. I saw you added mode into the AllNullQueriesTest, but somehow I didn't find where the null intermediate result is handled

The change I made there was to create one test per query instead of executing all in the same test method. The only reason to apply the change is to improve the UX in the case there is at least one failure. Before only the first error was reported and once you fix the issue there you had to run the tests again to check if there were other failures. Now all failures are reported at the same time.

I've modified the code long ago, so I don't remember the exact changes, but reading the diff it doesn't look I've changed anything related to null intermediate result.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java (hunk 8 above)
Comment (gortiz) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:504]: _this map_ = `intermediateResult`?

In that case, no. `intermediateResult` is created in `extractAggregationResult(AggregationResultHolder)`, which calls `extractIntermediateResult` which returns not null.

Comment (gortiz) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:504]: I'm adding a new test to show the result when all stored data is null

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java (hunk 7 above)
Comment (gortiz) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:146]: It is not used here, but it is being used in other PRs. Given I have so many PRs sharing the same base code, I've been merging them whenever I found an issue in the other PRs. For example, this method is being used in https://github.com/apache/pinot/pull/12271.

I guess I can remove it from here if I carefully merge them later

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java (hunk 6 above)
Comment (gortiz) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java]: IntIterator.close is abstract in RoaringBitmaps, so we have to implement it. Anyway, I've just remove this code to introduce it in the percentille PR, which is the one that actually uses it

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java (hunk 9 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java:259]: I think it is the same case than `orNull` int iterator. I'm going to remove it for now

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java [hunk 10]
```diff
@@ -263,11 +264,14 @@ public void aggregate(int length, AggregationResultHolder aggregationResultHolde
     // For dictionary-encoded expression, store dictionary ids into the dictId map
     Dictionary dictionary = blockValSet.getDictionary();
     if (dictionary != null) {
-      int[] dictIds = blockValSet.getDictionaryIdsSV();
+
       Int2IntOpenHashMap dictIdValueMap = getDictIdCountMap(aggregationResultHolder, dictionary);
-      for (int i = 0; i < length; i++) {
-        dictIdValueMap.merge(dictIds[i], 1, Integer::sum);
-      }
+      int[] dicts = blockValSet.getDictionaryIdsSV();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:269]: lol, you really hate the name `dictIds`

**Review by Jackie-Jiang — APPROVED** [review]
LGTM
